Async session creation

christian-bromann opened the pull request:
As a cloud provider starting a WebDriver session can take a while for arbitrary reasons. Long session creation times are problematic for users as request libraries tend to time out after a while. Currently workarounds are used where you would respond with a 302 redirect, redirecting to the same resource. Having a way to asynchronously create a session would help to get rid of such hacky workarounds.

This patch was [original authored](https://gist.github.com/shs96c/108f5313eae54b94658ee018e37926d2) by @shs96c during TPAC 2018. I would love to finally pick this up and bring it into the spec.

At a high level these would be the step to create an asynchronous session:

1. Send a request with same payload as for synchronous session requests to `[POST] /session/async`
   Response payload would be as follows:
   ```json
   {
       "sessionCreationJobId": "6fcd0780e995bd50fedd330cd49d0531",
       "status": "queueing",
       "timeout": 600000,
       "requestedTime": 1601990778990,
       "events": [],
       "data": {}
   }
   ```
1. Regularly check for the status of the session creation job by making requests to `[GET] /session/async/6fcd0780e995bd50fedd330cd49d0531`
   Response payload once job finished loading would be
   ```json
   {
       "sessionCreationJobId": "6fcd0780e995bd50fedd330cd49d0531",
       "status": "created",
       "timeout": 600000,
       "requestedTime": 1601990778990,
       "endedTime": 1601990878990,
       "events": [{ eventId: "eventId-1" }, { eventId: "eventId-2" }, { eventId: "eventId-3" }, { eventId: "eventId-4" }, ...],
       "data": {}
   }
   ```
1. If there is interest introspecting events happen while creating the job the user can query for an event id via `[GET] /session/async/6fcd0780e995bd50fedd330cd49d0531/eventId-1`
   Response payload for this request would contain
   ```json
   {
       "eventId": "event-1",
       "time": 1601990778990,
       "type": "progress",
       "message": "...",
       "data": "...",
       "stacktrace": "..."
   }
   ```

Notes from Simon when he drafted this with my comments:

> Should session creation progress be represented using “events” or “logs”?

Given that session creation progress is always connected to a point in time I would favor `events` as representation for it.

> Should session creation jobs have multiple timeouts, for example overall job timeout and wait timeout?

One timeout for the maximal amount of time it should take to create a job is fine. Overall job timeout is set as session timeout.

> Should we require representation of session creation attempts? (Compare [AWS Batch Describe Jobs](https://docs.aws.amazon.com/en_pv/batch/latest/APIReference/API_DescribeJobs.html))

IMO attempts can be easier represented in the bindings implementation rather than having to spec this into the protocol.

> Should we use a common suffix like <code>Time</code> for names of properties representing times? (e.g. the [Appium event timings API](http://appium.io/docs/en/advanced-concepts/event-timings/) uses the suffix <code>Time</code> but represents times using “JS timestamps”, presumably milliseconds as provided by [`Date.now`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/now), the [AWS Device Farm Get Job endpoint](https://docs.aws.amazon.com/en_pv/devicefarm/latest/APIReference/API_GetJob.html) uses no suffix, for example "<code>started</code>", the [SauceLabs Get Tunnel endpoint](https://wiki.saucelabs.com/display/DOCS/Tunnel+Methods#TunnelMethods-GetTunnel) uses no suffix, for example "<code>last_connected</code>". If we do use a suffix, should we specify the units, for example <code>endedTimeMillis</code>?

I propose how it is currently spec'ed with `endedTime` and `requestedTime`.

## Questions:
- @shs96c it is not quite clear for me how the user receives the `sessionId` after the user has fetched a created async job, what am I missing?


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/christian-bromann/webdriver/pull/1552.html" title="Last updated on Oct 28, 2020, 3:47 PM UTC (afd7aee)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webdriver/1552/99f7ced...christian-bromann:afd7aee.html" title="Last updated on Oct 28, 2020, 3:47 PM UTC (afd7aee)">Diff</a>

shs96c commented on index.html:
```diff
@@ -194,6 +194,10 @@ <h2>Terminology</h2>
 An <dfn>integer</dfn> is a <a>Number</a> that is unchanged
 under the <a>ToInteger</a> operation.
 
+<p>A <dfn>millisecond timestamp</dfn> is an <a>Integer</a> that
```
`Date.prototype.toJSON` delegates to [Date.prototype.toISOString](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString) I suggest we lean into that, rather than epoch millis

shs96c commented on index.html:
```diff
@@ -1144,13 +1172,29 @@ <h3>Errors</h3>
+  <td>404
+  <td><code>invalid session creation job id</code>
+  <td>Occurs if the given <a>session creation job id</a> is not in the list of <a>session creation jobs</a>,
+   or is in the list but has a <a data-lt="session creation status">status</a> of
```
While I can see the reasoning for this, it might be useful to be able to get the `cancelled` or `failed` event for inspection.

shs96c commented on index.html:
```diff
@@ -2139,8 +2190,203 @@ <h2>Sessions</h2>
+<td>"<code>requestedTime</code>"
+<td>
+<td>
+<td>When the session creation was requested as a <a>millisecond timestamp</a>.
```
As a `Date` would be more appropriate. Implementations would be free to store this as a millisecond timestamp.

shs96c commented on index.html:
```diff
@@ -2139,8 +2190,203 @@ <h2>Sessions</h2>
+<td>"<code>endedTime</code>"
+<td>✓
+<td>
+<td>When the job ended in success or failure as a <a>millisecond timestamp</a>.
```
As above.

shs96c commented on index.html:
```diff
@@ -2139,8 +2190,203 @@ <h2>Sessions</h2>
+<td>"<code>time</code>"
+<td>
+<td>
+<td>Event time as a <a>millisecond timestamp</a>.
```
`Date`

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+      <dt>"<code>capabilities</code>"
+      <dd>The capabilities of the <a>current session</a>.
+
+      <dt>"<code>capabilitiesRequest</code>"
```
Drop `capabilitiesRequest`

The shape of this object should match the shape of the object returned by `New Session`. The local end is aware of the capabilities that were requested, and (more importantly) this might be absolutely gigantic if someone has include one or more browser profiles.

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+    </dl>
+  </dl>
+
+  <li><p>Return success with data <var>body</var>.
```
You need to add a response for what to do if the `session id` is unknown.

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+<p>The <a>remote end</a> steps are:
+
+<ol>
+ <li><p>Let <var>session creation requested time</var> be the current time
```
If we use dates instead of millisecond timestamps, you can drop the next line.

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+  <ol>
+   <li><p>If <a>maximum timeout for async session creation</a> is an <a>integer</a>,
+    and <var>timeout</var> is undefined,
+    let <var>timeout</var> be <a>min</a>(600,000, <a>maximum timeout for async session creation</a>).
```
We may want implementations to be able to set a lower session creation timeout. While a minute is a sensible default for Sauce, on a local machine with a local driver, I'd expect async session creation to be significantly faster.

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+   <li><p>If <var>timeout</var> is <a><code>null</code></a>,
+    let <var>timeout</var> be <a>maximum timeout for async session creation</a>.
+   <li><p>If <a>maximum timeout for async session creation</a> is an <a>integer</a>,
+    and less than <var>timeout</var>,
```
It'll be better to use `maximum timeout for async session creation`: presumably this occurs when the local end sets some ridiculous timeout and the remote end has something sensible in place.

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+    <dd>An empty JSON <a>List</a>.
+
+     <dt>"<code>data</code>"
+     <dd>A JSON <a>Object</a> with implementation-specific data.
```
It's not clear what this `data` is meant to be used for. A note explaining this would be useful.

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+       <li><p>If <var>event</var> equals <var>event id</var>,
+        set <var>found</var> to true.
+
+       <li><p>Prepend <var>event object</var> to <var>filtered events</var>.
```
We only ever collect a single event from this loop. Why do we maintain the list?

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+      </ol>
+
+      <li><p>If <var>found</var> is false,
+       return <a>error</a> with <a>error code</a> <a>invalid event id</a>.
```
I think it's legit to just send the empty list as it just means that there have been no events, right?

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+
+ <li><p>If <var>session creation job</var> has a <a data-lt="session creation status">status</a> of
+  "<code>cancelled</code>" or "<code>failed</code>",
+  return <a>error</a> with <a>error code</a> <a>invalid session creation job id</a>.
```
Seems like the user-observable response would be the same as having deleted the job, no? If that's the case, should we consider returning "success with data null"?

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+   <li>Set the <a data-lt="session creation status">status</a> to "<code>cancelled</code>".
+
+   <li>If <var>session creation job</var>’s job <a>session ID</a> references an <a>active session</a>,
+   schedule a background job to <a>try</a> to <a>close the session</a>.
```
"schedule a background job" is an implementation-specific detail. I can imagine a local driver doing this serially.

shs96c commented on index.html:
```diff
@@ -2449,6 +2695,254 @@ <h3 id=new-session><dfn data-lt="new sessions|creating a new session">New Sessio
+  <ol>
+   <li>Set <var>session creation job</var>’s <a data-lt="session creation status">status</a> to "<code>cancelled</code>".
+   <li>Set <var>session creation job</var>’s <a data-lt="session creation ended time">ended time</a> to the current <a>millisecond timestamp</a>.
+   <li>Schedule a background job to cancel the session creation.
```
Again, avoid "background job". We also don't seem to remove the `session creation job` from the queue....

shs96c reviewed (changes requested):
This looks useful, and a good start. Thank you for picking this up.

christian-bromann commented on index.html:
```diff
@@ -2139,8 +2190,203 @@ <h2>Sessions</h2>
+<td>"<code>data</code>"
+<td>✓
+<td>
+<td>Object containing implementation-defined data associated with this job.
```
We should add a not that this will be not further specified as it is implementation specific.

AutomatedTester commented:
I can't see a link to tests so adding requires test label

jgraham commented:
I'd like to propose we start a new spec for this. The concern is that this doesn't seem like something browser vendors are likely to implement directly but something we could standardise with the implementation target being intermediary nodes. I know that's overhead, but unless there's a compelling reason that we actually need this in browser drivers I think it makes sense to seperate out so it's clear who the feature is aimed at.

shs96c commented:
I'm fine with us speccing this out for intermediate nodes alone, but it should be specced out.

christian-bromann commented:
@shs96c I addressed the comments and currently have only 2 remaining issue that I need input for:

1. How is the WebDriver use suppose to receive the session id from once the session creation job moved to `status` `created`?
2. In your original draft you defined an endpoint to fetch the session creation job (`/session/async/{session creation job id}`) as well as specific events via `/session/async/{session creation job id}/{event}` where `event` is an event id that a user can receive from the events property of the session creation job. According to the spec it would only find one event that is matching the event id provided by the url path, you commented:

    > I think it's legit to just send the empty list as it just means that there have been no events, right?

    which assumes it is able to fetch multiple events. I think the reasoning behind not allowing to just fetch all events is the same we have with logs however I don't think implementers will have as many job creation events as logs on a page. How about instead of fetching a specific event, to just have `/session/async/{session creation job id}/events` that fetches all events happening during job creation.

What do you all think?

christian-bromann commented:
Closing in favor of #1561
